Complex Stellar Populations

maxbriel opened the pull request:
This submodule introduces the possibility to create Complex Stellar Populations with BPASS data. 

The two main modules are `csp.eventrate` and `csp.spectra`
`csp.eventrate` contains the `CSPEventRate` class, which can be used to calculate event rates from a given Stellar Formation History and Metallicity evolution. 

The following methods are available in this class:
- `at_time`: calculates the event rates at a given lookback time.
- `over_time`: calculates the event rates over a lookback time (binned).
- `grid_at_time`: alternative input. Takes a grid of points instead of a function. Does the same as `at_time`.
- `grid_over_time`: alternative input. Takes a grid of points instead of a function. Does the same as `over_time`.

`csp.spectra` contains the `CSPSpectra` class, which can be used to calculate synthetic spectra for the given population.
The same methods are available in this class as `CSPEventRate`,  but provide spectra rather than even rates.

`csp.sfh` contains a class with which a Stellar Formation History (SFH) can be made. A collection of common parametric SFH are available (see documentation), but a custom one can also be given.

As support for these high level functions, a collection of utility and helper functions has been added. These can be found in `csp.utils`. Some these are optimised for performance (numba JIT compiled) and are not user friendly.

maxbriel committed:
Created CSP archistecture and basic sfh class

The complex stellar population subpackage is meant to combine bpass 
variables with a stelalr formation history to generate more complex 
populations. To this end, we're implementing a stellar formation history 
class. The basic SFH class for custom SFH is implemented using scipy 
splines. This basic class is also tested.

maxbriel committed:
Added variables to constants

maxbriel committed:
implements utility functions & new test files

maxbriel committed:
Implements event rate calculation over time

maxbriel committed:
BPASS data_files for testing file_loading

maxbriel committed:
Adds `at_time` function to the event rate class

This method allows for the calculation of the event rate at a given lookback time. It uses the BPASS binning for this. 

This method is much faster, but suffers from a lack of accuracy.

maxbriel committed:
Adds docstrings and extra tests

maxbriel committed:
Implements calculate spectrum at_time and over_time functions

Both functions are implemented for the BPASS spectra. 
However, the method is similar to the BPASS event rate calculation, which is sub optimal for the spectra calculation. It takes an extremely long time to do the calculation. A for loop of 100000 iterations is present, because each wavelength is seen as it's own "event type". It should be possible to remove this by implementing a function to take the separate wavelengths as a single unit. It would be interesting so see the influence of putting the for loop within a numba function, but thinking about restructuring the function is probably better. 

Including a "caching" of the BPASS spectra into a pickled DataFrame. 



Additional test files are required for this to run.

maxbriel committed:
Add normalisation function for the BPASS spectrum

HeloiseS committed:
fixed hrd plot

HeloiseS committed:
adding csp folder

HeloiseS committed:
fixed merge conflict

HeloiseS committed:
minor changes

maxbriel committed:
Moved tests to CSP folder + fix in function name _type_check

HeloiseS committed:
added a hokitype error

HeloiseS committed:
bringing it latest max update

HeloiseS committed:
Added some code review for Max and minor stylistic changes

HeloiseS committed:
Added some code review for Max and minor stylistic changes

maxbriel committed:
Updated the tests

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

HeloiseS committed:
bla

maxbriel committed:
Adjust tests and docstrings

maxbriel committed:
Adjust eventrate functions to take lists of functions

Instead of requiring a scipy interpolated spline as input. These functions now take a list of functions as input for the stellar formation and metallicity history. 

This adjustment has been made by using `numpy.interp` instead of scipy.interpolate. Furthermore, instead of scipy's spline integration `numpy.trapz` is now used to calculate the mass per bin.

HeloiseS committed:
fix merge comflict

maxbriel committed:
Disabled spectra test

HeloiseS committed:
Merge update from max 'csp' of https://github.com/HeloiseS/hoki into csp

HeloiseS committed:
Added parametric star formation histories to SFH and refactored some variable names

HeloiseS committed:
fixing CI

maxbriel committed:
Auto stash before merge of "csp" and "official_hoki/csp"

maxbriel committed:
Merge remote-tracking branch 'official_hoki/csp' into csp

maxbriel committed:
upload files for remote work

maxbriel committed:
Improve testing and functions

- Adds BPASS IMFS to the constants
- Adds check to BPASS IMFS
- Adds test for `data_compiler.SpectraData`
- Reduces data usage for rate tests

maxbriel committed:
Add SpectraCompiler test

Uses `unittest.mock` to allow for a smaller DataFrame to be used.

maxbriel committed:
Adds extensive tests for the CSP package

maxbriel committed:
Update SpectraCompiler

The compiler now outputs the BPASS spectra as a 3D `numpy.ndarray` this minimises storage and allows for faster and less memory hungry operations than the MultiIndex pandas DataFrame. 

The tests are updated accordingly and using unittest.mock a single spectra file is use. (This could be simulated too)

maxbriel committed:
Documentation update + SFH input

The documentation now shows the size and shape of each input and output variable. 

The `_over_time` and `_at_time` for `CSPSpectra` and `CSPEventRate` are now able to take the `SFH` object as an input.

maxbriel committed:
Removed CSP from utils due to cyclical import

The CSP class in `utils` was importing the SFH class, which was importing the `utils` module. Therefore, failing importing. Thus, the CSP class has been moved to a separate file

maxbriel committed:
Updated setup.cfg to include csp_test_data subfolder

maxbriel committed:
Removes leftover print statement + prints loading precompiled failed

maxbriel committed:
Correct metallicity mid-points

HeloiseS committed:
Added the parametric star formation histories as public functions

maxbriel committed:
at_time new default sample rate + load functions moved to hoki.load

at_time now uses 1000 as the default sample rate. 
The BPASS bins can still be used is a negative number is given. 
The `load_spectra` and `load_rates` have been moved to `hoki.load` and renamed to `all_spectra` and `all_rates`, respectively.
tests have been adjusted accordingly

maxbriel committed:
Fixed typo

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

maxbriel committed:
Removes cyclical import

maxbriel committed:
Adjusted tests that use SpectraCompiler

maxbriel committed:
Getting tox to work

maxbriel committed:
private optimised mass_per_time_bin calculation

Implements 2 new functions:
- A custom numba trapezoidal function
- A optimised mass_per_bin function for non-function-based data

maxbriel committed:
Refactored Event Rate calculation from array data

The original event rate calculation took functions as in input. This becomes cumbersome when you have data at specific time values. This calculation is done by taking a 2D matrix with time points on one axis and BPASS metallicities on the other. The values of the matrix describes the SFR at that moment in time. 

The calculate_rate_over_time function has been adjusted to use this new function.

maxbriel committed:
Removed dependency and added more doc information

findesgh committed:
BUG: pass on kwarg to spectra compiler.

findesgh committed:
Merge branch 'csp' of https://github.com/HeloiseS/hoki into csp

maxbriel committed:
Merge pull request #58 from findesgh/csp

Small bugfix in `all_spectra`
The binary `kwarg` was not passed on to `SpectraCompiler` from `load.all_spectra`.

maxbriel committed:
Fix SpectraCompiler

Commit:
BUG: use np instead of pd in SpectraCompiler.

Commit:
BUG: fix index and make test pass.

Commit:
BUG: fix another regression bug.

Commit:
Merge branch 'csp' of https://github.com/HeloiseS/hoki into csp_fix2

Commit:
Fix further regression bug.

maxbriel committed:
Clean up loading module

maxbriel committed:
Updates mass per bin calculation for vectorized funciton input

mass_per_bin now takes a normal python callable and a vectorized function as input. 
The latter boosts performance significantly

maxbriel committed:
Renames CSPEventRate functions and adds grid functions

- CSPEventRate functions are now called:
    - calculate_rate_over_time -> over_time
    - calculate_rate_at_time -> at_time
new functions: 
    - grid_over_time: calculates the rates from a 2D SFH (per BPASS metallicity) from  time_points to time bins
     - grid_at_time: calculate the event rate at a specific moment in lookback time from a 2D SFH grid

maxbriel committed:
np.zeros to np.empty where possible

maxbriel committed:
Merge pull request #60 from findesgh/csp_fix2

* BUG: use np instead of pd in SpectraCompiler.
* BUG: fix index and make test pass.
* BUG: fix another regression bug.
* Fix further regression bug.

Authored-by: Martin Glatzle <mglatzle@mpa-garching.mpg.de>

maxbriel committed:
Merge remote-tracking branch 'origin/csp' into csp

maxbriel committed:
Fixed incorrect test of at_time and over_time

maxbriel committed:
Set default sample rate for at_time

Also includes some docstring updates

maxbriel committed:
Adds `cache` parameter to over_time_spectrum

maxbriel committed:
Adds grid functions for the spectra calculation

CSPSpectra now has the following functions: 
Public: 
- at_time: at time with function input
- over_time: over lb time with function input
- grid_at_time: at time with SFH grid input
- grid_over_time: over lb time with SFH grid input
Private: 
- grid_rate_calculator_at_time
- grid_rate_calculator_over_time

The grid_rate_calculators_* are numba functions that  have the grid_* around them as wrappers for the input and output

maxbriel committed:
Adds return_time_edges to grid_over_time & bug fix sfh vectorisation

maxbriel committed:
Bug fix in nr_time_bins (grid_over_time functions)

Instead of N bins being created, N-1 bin was being created.

maxbriel committed:
Docs string cleanup

maxbriel committed:
Merge branch 'dev1.6' into csp

maxbriel committed:
Solve cyclical import error

maxbriel committed:
Fix tox and remove corner plot

HeloiseS commented on hoki/csp/tests/test_eventrate.py:
```diff
@@ -0,0 +1,180 @@
+            "Correct input is not taken."
+
+        assert np.isclose(self.CSP.at_time(sfh_fnc,
+                                                          Z_fnc,
```
what happened here?

maxbriel replied:
If you mean the indenting, that's caused by the change of the function name to a shorter `at_time` instead of `eventrate_at_time`.

HeloiseS commented on hoki/data_compilers.py:
```diff
@@ -1,19 +1,23 @@
-bpass_input_z_list = ['zem5','zem4', 'z001', 'z002', 'z003', 'z004', 'z006',
-                       'z008', 'z014', 'z010', 'z020','z030','z040']
-
+bpass_input_z_list = ["z"+i for i in BPASS_METALLICITIES]
```
is there a reason the non numerical BPASS metalicities can't have the z in their string directly?

maxbriel replied:
No specific reason. More that z is constant between all metallicities, so there's no specific reason to have it be part of the list. Here I noticed that `bpass_input_z_list` is compared to user input, and I did not want to change that/didn't know where the user input was described. Therefore, I kept it like this. But we can change `BPASS_METALLICITIES` to contain the `z`. Although I'm not sure it should be part of it.

HeloiseS replied:
I'm just thinking that having two lists is redundant is all

maxbriel replied:
What do we want? With or without? 
With means I go search for when I'm using it. 
Without means changing the tutorial for the DataCompiler.

HeloiseS replied:
Let's meet at 10.50 on the zoom channel for the group meeting and find the best solution, will be easier than with comments :)

maxbriel replied:
Renamed to BPASS_METALLICITIES and now included "z"

HeloiseS commented on hoki/data_compilers.py:
```diff
@@ -207,14 +284,11 @@ def _print_welcome():
 
     print(
-        "\nNOTE: The progress bar doesn't move smoothly - it might accelerate or slow down - it'dc perfectly normal :D")
+      "\nNOTE: The progress bar doesn't move smoothly - it might accelerate or slow down - it's perfectly normal :D")
```
hahah thanks that was pissing me off but i kept forgetting to fix it

HeloiseS commented on hoki/load.py:
```diff
@@ -410,11 +411,173 @@ def _UV_nebular_lines(path):
+    return rates.swaplevel(0, 1, axis=1)
+
+
+def all_spectra(data_path, imf, binary=True):
```
this name is too vague - at face value it doesn't tell your user what exactly is being loaded and that a file is being created automatically. we can talk about this

maxbriel replied:
`spectra_all_z_cached`?

maxbriel replied:
named `spectra_all_z` in commit 7801d16

HeloiseS commented on hoki/load.py:
```diff
@@ -410,11 +411,173 @@ def _UV_nebular_lines(path):
+#  BPASS Load over all metallicity  #
+#####################################
+
+def all_rates(data_path, imf, binary=True):
```
rates_all_z

HeloiseS commented on hoki/load.py:
```diff
@@ -410,11 +411,173 @@ def _UV_nebular_lines(path):
+    # load supernova count files
+    for num, metallicity in enumerate(BPASS_METALLICITIES):
+        data = model_output(
+            f"{data_path}/supernova-{star}-{imf}.z{metallicity}.dat"
```
what if the path is incorrect? If they haven't got all the folders for all the metallicities for example?

maxbriel replied:
This should be caught by the `model_output`, which checks for the existence of the file. So it will return an assert error, if a metallicity file is missing.
More problematic is the `all_spectra`, which currently does not check for the existence of the files. 

Edit: `load_spectra` to `all_spectra`

HeloiseS replied:
Ah yes true true. And I forgot that load spectra didn't do that. Bug oversight on my part. Can you make that an issue? I'll fix it soon.

maxbriel replied:
That's my issue ;) It's `all_spectra` in my code that doesn't check for it. `model_output`'s spectra load does perform a proper check.

maxbriel replied:
Check implemented in commit 7801d16

HeloiseS reviewed (changes requested):
This is fantastic work I can't wait to use it - sorry it took me a few days to look at it

I have a few minor comments and questions mostly regarding the new functions added to load.

maxbriel committed:
File check, combined data_compiler tests, renames func+variables

renamed variables:
- BPASS_METALLICITIES. now with `z`
renamed functions:
- all_rates ->  rates_all_z
- all_spectra -> spectra_all_z
